File: design-circular-queue.cpp
// https://leetcode.com/problems/design-circular-queue/

class MyCircularQueue {
    unique_ptr<int[]> data_;
    int capacity_, size_, frontIdx_;

public:
    /** Initialize your data structure here. Set the size of the queue to be k. */
    MyCircularQueue(int k)
        : data_(new int[k]), capacity_(k), size_(0), frontIdx_(0) {}

    /** Insert an element into the circular queue. Return true if the operation is successful. */
    bool enQueue(int value)
    {
        if (isFull()) {
            return false;
        }

        data_[(frontIdx_ + size_++) % capacity_] = value;
        return true;
    }

    /** Delete an element from the circular queue. Return true if the operation is successful. */
    bool deQueue()
    {
        if (isEmpty()) {
            return false;
        }

        frontIdx_ = --size_ > 0 ? (frontIdx_ + 1) % capacity_ : 0;
        return true;
    }

    /** Get the front item from the queue. */
    int Front()
    {
        return !isEmpty() ? data_[frontIdx_] : -1;
    }

    /** Get the last item from the queue. */
    int Rear()
    {
        return !isEmpty() ? data_[(frontIdx_ + size_ - 1) % capacity_] : -1;
    }

    /** Checks whether the circular queue is empty or not. */
    bool isEmpty()
    {
        return size_ == 0;
    }

    /** Checks whether the circular queue is full or not. */
    bool isFull()
    {
        return size_ == capacity_;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */

File: leetcode/medium/design-circular-queue.rs
// https://leetcode.com/problems/design-circular-queue
//
// Complexity: runtime O(1) all operations (except O(n) construction), space O(n).

#[derive(Default)]
struct MyCircularQueue {
    buf: Vec<i32>,
    front: usize,
    len: usize,
}

impl MyCircularQueue {
    fn new(k: i32) -> Self {
        Self {
            buf: vec![0; k as _],
            ..Default::default()
        }
    }

    fn en_queue(&mut self, value: i32) -> bool {
        if self.is_full() {
            false
        } else {
            let cap = self.buf.len();
            self.len += 1;
            self.buf[(self.front + self.len - 1) % cap] = value;
            true
        }
    }

    fn de_queue(&mut self) -> bool {
        if self.is_empty() {
            false
        } else {
            self.len -= 1;
            self.front = (self.front + 1) % self.buf.len();
            true
        }
    }

    fn front(&self) -> i32 {
        if self.is_empty() {
            -1
        } else {
            self.buf[self.front]
        }
    }

    fn rear(&self) -> i32 {
        if self.is_empty() {
            -1
        } else {
            self.buf[(self.front + self.len - 1) % self.buf.len()]
        }
    }

    fn is_empty(&self) -> bool {
        self.len == 0
    }

    fn is_full(&self) -> bool {
        self.len == self.buf.len()
    }
}
